File: web/src/components/sections/__tests__/CalendarStripSchemaDecode.test.ts
import { describe, expect, it } from 'vitest';
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';
import type { Section } from '@sdui/models';
import { mapCalendarStrip } from '../../../adapters/sectionUiAdapters';

function loadFixtureSection(): Section {
  const fixturePath = resolve(__dirname, '../../../../../schema/examples/calendar-strip.json');
  return JSON.parse(readFileSync(fixturePath, 'utf-8')) as Section;
}

describe('CalendarStrip schema decode contract', () => {
  it('decodes fixture with all expected fields', () => {
    const section = loadFixtureSection();
    const model = mapCalendarStrip(section);

    expect(section.id).toBe('server:games-calendar~type=CalendarStrip');
    expect(section.type).toBe('CalendarStrip');
    expect(section.contentSourceId).toBe('server:games-calendar');
    expect(section.analyticsId).toBe('games_calendar_strip');
    expect(section.accessibility?.label).toBe('Games date picker');

    expect(model).not.toBeNull();
    expect(model?.stateKey).toBe('games_selected_date');
    expect(model?.selectedDate).toBe('2026-05-25');
    expect(model?.defaultDate).toBe('2026-05-25');
    expect(model?.minDate).toBe('2025-10-01');
    expect(model?.maxDate).toBe('2026-06-30');
    expect(model?.onDateSelected.trigger).toBe('onActivate');
    expect(model?.onDateSelected.type).toBe('refresh');
    expect(model?.onDateSelected.endpoint).toBe('/v1/sdui/screen/refresh/games');
    expect(model?.onDateSelected.paramBindings?.date).toBe('{{games_selected_date}}');
  });

  it('fails decode when defaultDate is omitted', () => {
    const section = {
      id: 'server:games-calendar~type=CalendarStrip',
      type: 'CalendarStrip',
      data: {
        stateKey: 'games_selected_date',
        selectedDate: '2026-05-25',
        onDateSelected: {
          trigger: 'onActivate',
          type: 'refresh',
          endpoint: '/v1/sdui/screen/refresh/games',
          paramBindings: { date: '{{games_selected_date}}' },
        },
      },
    } as Section;

    expect(mapCalendarStrip(section)).toBeNull();
  });

  it('succeeds decode when minDate and maxDate are omitted', () => {
    const section = {
      id: 'server:games-calendar~type=CalendarStrip',
      type: 'CalendarStrip',
      data: {
        stateKey: 'games_selected_date',
        selectedDate: '2026-05-25',
        defaultDate: '2026-05-25',
        onDateSelected: {
          trigger: 'onActivate',
          type: 'refresh',
          endpoint: '/v1/sdui/screen/refresh/games',
          paramBindings: { date: '{{games_selected_date}}' },
        },
      },
    } as Section;

    const model = mapCalendarStrip(section);
    expect(model).not.toBeNull();
    expect(model?.minDate).toBeUndefined();
    expect(model?.maxDate).toBeUndefined();
  });

  it('fails decode when onDateSelected is an array', () => {
    const section = {
      id: 'server:games-calendar~type=CalendarStrip',
      type: 'CalendarStrip',
      data: {
        stateKey: 'games_selected_date',
        selectedDate: '2026-05-25',
        defaultDate: '2026-05-25',
        onDateSelected: [
          {
            trigger: 'onActivate',
            type: 'refresh',
            endpoint: '/v1/sdui/screen/refresh/games',
            paramBindings: { date: '{{games_selected_date}}' },
          },
        ],
      },
    } as unknown as Section;

    expect(mapCalendarStrip(section)).toBeNull();
  });
});

File: android/sdui-core/src/test/java/com/nba/sdui/core/CalendarStripSchemaDecodeTest.kt
package com.nba.sdui.core

import com.fasterxml.jackson.module.kotlin.readValue
import com.nba.sdui.core.models.generated.Section
import com.nba.sdui.core.models.generated.mapper
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test
import java.io.File

class CalendarStripSchemaDecodeTest {

    @Test
    fun `calendar-strip fixture decodes with expected fields`() {
        val section: Section = mapper.readValue(loadExample("calendar-strip.json"))

        assertEquals("server:games-calendar~type=CalendarStrip", section.id)
        assertEquals("CalendarStrip", section.type)
        assertEquals("server:games-calendar", section.contentSourceID)
        assertEquals("games_calendar_strip", section.analyticsID)
        assertEquals("Games date picker", section.accessibility?.label)

        val data = section.data
        assertNotNull(data)
        assertEquals("games_selected_date", data?.stateKey)
        assertEquals("2026-05-25", data?.selectedDate)
        assertEquals("2026-05-25", data?.defaultDate)
        assertEquals("2025-10-01", data?.minDate)
        assertEquals("2026-06-30", data?.maxDate)
        assertEquals("onActivate", data?.onDateSelected?.trigger?.value)
        assertEquals("refresh", data?.onDateSelected?.type?.value)
        assertEquals("/v1/sdui/screen/games", data?.onDateSelected?.endpoint)
        assertEquals("{{games_selected_date}}", data?.onDateSelected?.paramBindings?.get("date"))
    }

    @Test
    fun `calendar-strip decode fails when defaultDate missing`() {
        val invalid = """
            {
              "id": "server:games-calendar~type=CalendarStrip",
              "type": "CalendarStrip",
              "data": {
                "stateKey": "games_selected_date",
                "selectedDate": "2026-05-25",
                "onDateSelected": {
                  "trigger": "onActivate",
                  "type": "refresh",
                  "endpoint": "/v1/sdui/screen/games",
                  "paramBindings": { "date": "{{games_selected_date}}" }
                }
              }
            }
        """.trimIndent()

        assertThrows(Exception::class.java) {
            mapper.readValue<Section>(invalid)
        }
    }

    @Test
    fun `calendar-strip decode succeeds when min and max omitted`() {
        val payload = """
            {
              "id": "server:games-calendar~type=CalendarStrip",
              "type": "CalendarStrip",
              "data": {
                "stateKey": "games_selected_date",
                "selectedDate": "2026-05-25",
                "defaultDate": "2026-05-25",
                "onDateSelected": {
                  "trigger": "onActivate",
                  "type": "refresh",
                  "endpoint": "/v1/sdui/screen/games",
                  "paramBindings": { "date": "{{games_selected_date}}" }
                }
              }
            }
        """.trimIndent()

        val section = mapper.readValue<Section>(payload)
        assertNull(section.data?.minDate)
        assertNull(section.data?.maxDate)
    }

    @Test
    fun `calendar-strip decode fails when onDateSelected is array`() {
        val invalid = """
            {
              "id": "server:games-calendar~type=CalendarStrip",
              "type": "CalendarStrip",
              "data": {
                "stateKey": "games_selected_date",
                "selectedDate": "2026-05-25",
                "defaultDate": "2026-05-25",
                "onDateSelected": [
                  {
                    "trigger": "onActivate",
                    "type": "refresh",
                    "endpoint": "/v1/sdui/screen/games",
                    "paramBindings": { "date": "{{games_selected_date}}" }
                  }
                ]
              }
            }
        """.trimIndent()

        assertThrows(Exception::class.java) {
            mapper.readValue<Section>(invalid)
        }
    }

    private fun loadExample(fileName: String): String {
        var current = File(System.getProperty("user.dir"))
        repeat(6) {
            val candidate = File(current, "schema/examples/$fileName")
            if (candidate.isFile) {
                return candidate.readText()
            }
            current = current.parentFile ?: return@repeat
        }
        error("schema/examples/$fileName not found from ${System.getProperty("user.dir")}")
    }
}
